docs/flags.md

```diff
@@ -0,0 +1,130 @@
+---
+title: Flags for Tiltfile Config Management
+layout: docs
+---
+
+
+This doc describes how to use flags to maintain configuration of a developer instance. This is especially useful for teams that have both enough services that there are multiple ways you might want to run it, and a designated owner of the developer experience (often a "DevEx" team or principal engineer). We'll use "you" to mean that DevEx engineer and "your user" to mean a developer at your org that uses Tilt.
+
+## Goal
+Users can start Tilt in a specific mode with a single Tilt command. For instance, `tilt up consumer` will start just the services for consumer users (vs other modes, like `tilt up enterprise`). If they later run `tilt up` it can stay in the selected mode, and they can change the mode while tilt is running with `tilt flags enterprise`.
+
+This doc describes how to describe your setup to Tilt to enable these shorthands. We aim to make common cases trivial, while also supporting escape hatches that enable arbitrarily-complex cases.
+
+In order, we'll cover:
+*) Examples that cover common cases, showing both the Tiltfiles and accompanying usage.
+*) Describe how this is implemented. These details, especially around persistence, will let you understand how to write tooling that interacts with this system.
+*) Future work (we see how it fits in but don't plan to implement until there's concrete demand).
+
+## Examples
+
+### Run only some services
+Your app has many services, call them A, B, C, and D. You want users to be able to run only some of them.
+
+and you want your users to be able to run only a subset of them. E.g., `tilt up` will run all services, but running `tilt up a b d` will run only the resources a, b, and d, ignoring c. Each run of Tilt will start from scratch.
+#### Command-Lines
+* `tilt up`: run all services
+* `tilt up a b d`: run A, B, and D but not C
+* `tilt flags a b`: change running Tilt to run A and B but neither C nor D.
+* `tilt up` (a second time): run all services
+
+#### Tiltfile
+```python
+flags.define_string_list("to-run", args=True, sticky=False)
+cfg = flags.parse()
+
+flags.exclude_all_resources()
+for r in cfg['to-run']:
+  flags.include_resource(r)
+```
+
+### Run a defined set of services
+Your app has many services, call them A, B, C, and D. You have defined modes you want users to be able to start (e.g. consumer vs. enterprise) and you don't want them to have to remember which service is necessary for which mode.
+
+You want to be able to write a doc that tells users to run `tilt up consumer` the first time and know that's enough that they'll be set up. Running Tilt subsequent times without arguments should default to the mode they already have set.
+
+#### Command-Lines
+* `tilt up`: run all services
+* `tilt up consumer`: run consumer services and record mode
+* `tilt up` (a second time): run consumer services
+
+#### Tiltfile
+```python
+flags.define_string_list("to-run", args=True)
```

Comment:
Is this analogous to an existing arg parsing library? would it make sense to start with something that could expand to, e.g., [argparse syntax](https://docs.python.org/3/library/argparse.html)? It wouldn't be great to invent our own syntax from scratch, either in terms of wasting time figuring things out ourselves, or in terms of potential lost preexisting user familiarity.

Reply:
In terms of how to define things in the Tiltfile or in terms of what's accepted on the command-line?

Reply:
the former, which might affect the latter